### ISSUE TITLE
How to make Keras Layer class to be immutable or is this a bug?

### ISSUE THREAD

**Issue by quaesitor-scientiam** [tensorflow/tensorflow#53738]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **from https://github.com/ahmedfgad/Mask-RCNN-TF2** with some other TF2 fixes

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  **Windows Pro**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:  **n/a**
- TensorFlow installed from (source or binary):  **via pip install**
- TensorFlow version (use command below):  **v2.7.0-rc1-69-gc256c071bb2 2.7.0**
- Python version:   **v3.7.3:ef4ec6ed12, Mar 25 2019, 22:22:05) [MSC v.1916 64 bit (AMD64)] on win32**
- Bazel version (if compiling from source):   **n/a**
- GCC/Compiler version (if compiling from source):   **n/a**
- CUDA/cuDNN version:   **cuda_11.2.r11.2/compiler.29373293_0**
- GPU model and memory:   **Quadro 2000 32GB**

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
**No windows version**

**Describe the current behavior**
The exception in the log below is symptom not the cause.  The actual cause is the variable used in the `set()` call in method `_insert_layers()` of the class `Functional`
```
    # Insert layers and update other layer attrs.
    layer_set = set(self._self_tracked_trackables) 
```
The `self._self_tracked_trackables` is a list of class instances of type `keras.engine.input_layer.InputLayer`.  Set requires the instances to be immutable and since it gives an error it must not be.  So  the question then becomes why pass something that  will not pass this requirement.   

We can find that `self._self_tracked_trackables` is initialized in the code below from method `_init_graph_network` of the same class which gets called when the model is initialized. It can be  seen that self.outputs is what drives the layers output of this `_map_graph_network` call.
```
    # Keep track of the network's nodes and layers.
    nodes, nodes_by_depth, layers, _ = _map_graph_network(
        self.inputs, self.outputs)
    self._network_nodes = nodes
    self._nodes_by_depth = nodes_by_depth
    self._self_tracked_trackables = layers
```
The model is defined as:
```
            model = KM.Model([input_image, input_image_meta, input_anchors],
                             [detections, mrcnn_class, mrcnn_bbox,
                                 mrcnn_mask, rpn_rois, rpn_class, rpn_bbox],
                             name='mask_rcnn')
```
Each of the outputs are derived from `keras.layers.Layer' class.  So how do  we make Layer to  be non-mutable to be able to use that `set()` call in the keras system?

**Describe the expected behavior**
No exception

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):  **no**
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
**Please see https://github.com/ahmedfgad/Mask-RCNN-TF2**

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
```
Exception has occurred: TypeError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
unhashable type: 'ListWrapper'
  File "D:\.venvs\maskrcnn\Lib\site-packages\tensorflow\python\training\tracking\data_structures.py", line 672, in __hash__
    raise TypeError("unhashable type: 'ListWrapper'")
  File "D:\.venvs\maskrcnn\Lib\site-packages\keras\engine\functional.py", line 851, in _insert_layers
    layer_set = set(self._self_tracked_trackables)
  File "D:\.venvs\maskrcnn\Lib\site-packages\keras\engine\functional.py", line 908, in _graph_network_add_loss
    self._insert_layers(new_layers, new_nodes)
  File "D:\.venvs\maskrcnn\Lib\site-packages\keras\engine\base_layer.py", line 1579, in add_loss
    self._graph_network_add_loss(symbolic_loss)
  File "D:\public\Mask-RCNN-TF2\mrcnn\model.py", line 2191, in compile
    self.keras_model.add_loss(loss)
  File "D:\public\Mask-RCNN-TF2\mrcnn\model.py", line 2379, in train
    self.compile(learning_rate, self.config.LEARNING_MOMENTUM)
  File "D:\public\Mask-RCNN-TF2\samples\shapes\debug_shapes.py", line 230, in <module>
    layers='heads')
  File "C:\Python\python37\Lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Python\python37\Lib\runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "C:\Python\python37\Lib\runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "C:\Python\python37\Lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Python\python37\Lib\runpy.py", line 193, in _run_module_as_main (Current frame)
    "__main__", mod_spec)
```

**Comment by sushreebarsa** [comment]
@quaesitor-scientiam 
Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see;
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)
Thank you!

**Comment by sushreebarsa** [comment]
@quaesitor-scientiam Thank you for the update!
Moving this issue to closed status as we will track this ticket [Keras #15897](https://github.com/keras-team/keras/issues/15897)
